src/transformers/models/gpt2/modeling_gpt2.py

```diff
@@ -395,6 +472,17 @@ def _init_weights(self, module):
             module.bias.data.zero_()
             module.weight.data.fill_(1.0)
 
+        # Reinitialize selected weights subject to the OpenAI GPT-2 Paper Scheme:
+        #   > A modified initialization which accounts for the accumulation on the residual path with model depth. Scale
+        #   > the weights of residual layers at initialization by a factor of 1/√N where N is the # of residual layers.
+        #   >   -- GPT-2 :: https://openai.com/blog/better-language-models/
+        #
+        # Reference (Megatron-LM): https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/model/gpt_model.py
+        for name, p in module.named_parameters():
```

Comment:
@siddk , this function is called from the parent class with `self.apply(self._init_weights)`, and then `named_parameters` is recursive too, so wouldn't the if statement below will trigger many times for the same weight tensor, for each layer of the onion above it? It seems like one fix would be to run this chunk of code once e.g. in post_init. Current behavior isn't a bug but is a bit wasteful in that it may be re-initializing the same tensor multiple times. And it looks a bit funny. Also would it be better to e.g. `name.endswith('c_proj.weight')`? Also not clear in the comment why 2 Layer Norms are relevant, did the comment maybe meant to say that there are two projections into the residual pathway per block?

Reply:
Or you can open a PR directly @karpathy so you get full credit for your find! :smiley:

Reply:
Haha, honestly my silly little proposed change may be as simple as: https://github.com/huggingface/transformers/commit/0637d69e83dbdc54a0f0fc0da20272c3818d97b7 . (I verified that this if statement triggers exactly 24 times for the simplest gpt2 model). Putting this in some kind of post_init would be much harder given how the code is laid out atm. Not super sure I should make a PR of it though 😅

Reply:
Ahah, one line of change for a 4x improvement, those are the best kinds of PRs!

Reply:
haha ok https://github.com/huggingface/transformers/pull/17877 . maybe I can do something more useful next ;)